pubsub/consumer.go

```diff
@@ -94,82 +85,138 @@ func (c *Consumer[Request, Response]) StreamName() string {
+func decrementMsgIdByOne(msgId string) string {
+	id, err := getUintParts(msgId)
+	if err != nil {
+		log.Error("Error decrementing start of XAutoClaim by one, defaulting to 0", "err", err)
+		return "0"
 	}
-}
-
-// heartBeat updates the heartBeat key indicating aliveness.
-func (c *Consumer[Request, Response]) heartBeat(ctx context.Context) {
-	if err := c.client.Set(ctx, c.heartBeatKey(), time.Now().UnixMilli(), 2*c.cfg.KeepAliveTimeout).Err(); err != nil {
-		l := log.Info
-		if ctx.Err() != nil {
-			l = log.Error
-		}
-		l("Updating heardbeat", "consumer", c.id, "error", err)
+	if id[1] > 0 {
+		return strconv.FormatUint(id[0], 10) + "-" + strconv.FormatUint(id[1]-1, 10)
+	} else if id[0] > 0 {
+		return strconv.FormatUint(id[0]-1, 10) + "-" + strconv.FormatUint(math.MaxUint64, 10)
+	} else {
+		log.Error("Error decrementing start of XAutoClaim by one, defaulting to 0", "err", err)
+		return "0"
 	}
 }
 
 // Consumer first checks it there exists pending message that is claimed by
 // unresponsive consumer, if not then reads from the stream.
-func (c *Consumer[Request, Response]) Consume(ctx context.Context) (*Message[Request], error) {
-	res, err := c.client.XReadGroup(ctx, &redis.XReadGroupArgs{
-		Group:    c.redisGroup,
-		Consumer: c.id,
-		// Receive only messages that were never delivered to any other consumer,
-		// that is, only new messages.
-		Streams: []string{c.redisStream, ">"},
-		Count:   1,
-		Block:   time.Millisecond, // 0 seems to block the read instead of immediately returning
-	}).Result()
-	if errors.Is(err, redis.Nil) {
-		return nil, nil
-	}
-	if err != nil {
-		return nil, fmt.Errorf("reading message for consumer: %q: %w", c.id, err)
+func (c *Consumer[Request, Response]) Consume(ctx context.Context) (*Message[Request], chan struct{}, error) {
+	// First try to XAUTOCLAIM, with start as a random messageID from PEL with MinIdle as IdletimeToAutoclaim
+	// this prioritizes processing PEL messages that have been waiting for more than IdletimeToAutoclaim duration
+	var messages []redis.XMessage
+	if pendingMsgs, err := c.client.XPendingExt(ctx, &redis.XPendingExtArgs{
+		Stream: c.redisStream,
+		Group:  c.redisGroup,
+		Start:  "-",
+		End:    "+",
+		Count:  math.MaxInt64,
+		Idle:   c.cfg.IdletimeToAutoclaim,
+	}).Result(); err != nil {
+		if !errors.Is(err, redis.Nil) {
+			log.Error("Error from XpendingExt in getting PEL for auto claim", "err", err, "penindlen", len(pendingMsgs))
+		}
+	} else if len(pendingMsgs) > 0 {
+		idx := rand.Intn(len(pendingMsgs))
+		messages, _, err = c.client.XAutoClaim(ctx, &redis.XAutoClaimArgs{
+			Group:    c.redisGroup,
+			Consumer: c.id,
+			MinIdle:  c.cfg.IdletimeToAutoclaim, // Minimum idle time for messages to claim (in milliseconds)
+			Stream:   c.redisStream,
+			Start:    decrementMsgIdByOne(pendingMsgs[idx].ID),
+			Count:    1,
+		}).Result()
+		if err != nil {
+			log.Error("error from xautoclaim", "err", err)
+		}
```

Comment:
I am FOR option 1 instead of 2 because that would be similar to trying to xautoclaim on every message from xpending, the reason we went with random pick is to have a balance between choosing idle pel entries and new messages with first preference given to idle pel entries.

retrying on idle pel entries again and again until success or xpending is exhausted might lead to unintended consequence of starving new messages. length of idle pel entries should anyway be less in number and shouldn't regularly increase - which again will only lead to starvation if rate of addition of entries to this idle pel entries is greater than rate of removal (which is equal to N/[avg processing time] where N is number of workers)